File: crates/compas-core/src/track.rs
use serde::{Deserialize, Serialize};

/// A streaming/metadata provider. `Local` is the first-class DJ path; the others
/// are control-only (see [`crate::SourceCapabilities`]).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MusicProvider {
    Local,
    Spotify,
    AppleMusic,
    SoundCloud,
}

/// Normalized track metadata shared across providers and the UI.
///
/// Ported and renamed from djvibebar's `TrackMetadata`. The original conflated
/// every provider id into a field named `spotify_id`; here it is `provider_id`
/// and the owning provider is explicit.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TrackMetadata {
    /// Stable local id (UUID) we assign on import/search.
    pub id: String,
    /// Provider that owns `provider_id` (Spotify track id, SoundCloud id, file path hash, …).
    pub provider: MusicProvider,
    /// The provider-native id. For `Local`, the absolute file path.
    pub provider_id: String,
    pub title: String,
    pub artist: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub album: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub artwork_url: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
    /// Detected/known tempo in beats-per-minute. `None` until analyzed.
    /// For streaming decks this is usually permanently `None` (no PCM to analyze,
    /// and Spotify's audio-features endpoint is no longer available to new apps).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub bpm: Option<f32>,
    /// Detected musical key (e.g. "8A" Camelot / "Am"). `None` until analyzed.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub musical_key: Option<String>,
}

impl TrackMetadata {
    /// Capabilities implied purely by the provider. Local files get full DSP;
    /// everything else is control-only.
    pub fn capabilities(&self) -> crate::SourceCapabilities {
        match self.provider {
            MusicProvider::Local => crate::SourceCapabilities::FULL_DSP,
            _ => crate::SourceCapabilities::PLAYBACK_ONLY,
        }
    }
}

File: apps/compas-dj/frontend/src/types/track.ts
// Mirrors `compas_core::TrackMetadata` / `SourceCapabilities`. Keep in sync with
// crates/compas-core. (A future step can codegen this from Rust via ts-rs/specta.)

export type MusicProvider = "local" | "spotify" | "apple_music" | "soundcloud";

export interface TrackMetadata {
  id: string;
  provider: MusicProvider;
  provider_id: string;
  title: string;
  artist: string;
  album?: string;
  artwork_url?: string;
  duration_ms?: number;
  bpm?: number;
  musical_key?: string;
}

export interface SourceCapabilities {
  full_dsp: boolean;
  provides_pcm: boolean;
  can_seek: boolean;
  can_vary_tempo: boolean;
}

/** Capabilities implied by the provider — the UI uses this to disable controls
 *  that cannot work for streaming decks (honest degradation, never faked DSP). */
export function capabilitiesFor(provider: MusicProvider): SourceCapabilities {
  if (provider === "local") {
    return { full_dsp: true, provides_pcm: true, can_seek: true, can_vary_tempo: true };
  }
  return { full_dsp: false, provides_pcm: false, can_seek: true, can_vary_tempo: false };
}
